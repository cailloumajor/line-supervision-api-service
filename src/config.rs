use std::path::PathBuf;
use std::str::FromStr;

use anyhow::{Context, Result};
use dotenv::dotenv;
use envconfig::Envconfig;
use url::Url;

#[derive(Clone)]
pub struct BaseUrl(Url);

impl FromStr for BaseUrl {
    type Err = url::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut base_path = s.to_string();
        if !base_path.ends_with('/') {
            base_path.push('/');
        }
        let base_url: Url = base_path.parse()?;

        Ok(Self(base_url))
    }
}

impl From<BaseUrl> for Url {
    fn from(base: BaseUrl) -> Self {
        base.0
    }
}

#[derive(Envconfig)]
pub struct Config {
    pub ui_customization_file: PathBuf,
    pub logo_file: PathBuf,
    pub influxdb_base_url: BaseUrl,
    pub influxdb_read_token: String,
    pub influxdb_org: String,
    pub influxdb_bucket: String,
}

impl Config {
    pub fn get() -> Result<Self> {
        dotenv().context("failed to get environment variables from .env file")?;
        let config = Config::init_from_env()
            .context("failed to get configuration from environment variables")?;
        Ok(config)
    }
}

#[cfg(test)]
mod tests {
    use test_case::test_case;

    use super::*;

    #[test_case("http://:8000" => panics "EmptyHost" ; "URL parse error")]
    #[test_case("http://host:8080" => "http://host:8080/" ; "appends slash")]
    #[test_case("http://host:8000/" => "http://host:8000/" ; "keeps slash")]
    #[test_case("http://host/path" => "http://host/path/" ; "appends slash to path")]
    #[test_case("http://host/path/" => "http://host/path/" ; "keeps slash after path")]
    fn base_url_from_str(string: &str) -> String {
        let base_url: BaseUrl = string.parse().unwrap();

        base_url.0.into()
    }

    #[test]
    fn url_from_base_url() {
        let url = Url::parse("http://host:80").unwrap();
        let base_url = BaseUrl(url.clone());
        let into_url: Url = base_url.into();
        assert_eq!(into_url, url)
    }
}
