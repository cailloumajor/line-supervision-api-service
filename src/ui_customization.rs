use anyhow::{Context, Result};
use serde_json::Value as JsonValue;
use std::fs;
use toml::Value as TomlValue;

use crate::config::Config;

fn camel_case(source: &str) -> String {
    let mut dest = String::with_capacity(source.len());
    let mut capitalize = false;
    for ch in source.chars() {
        if ch == ' ' || ch == '_' {
            capitalize = true;
        } else if capitalize {
            dest.push(ch.to_ascii_uppercase());
            capitalize = false;
        } else {
            dest.push(ch);
        }
    }
    dest[..1].to_ascii_lowercase() + &dest[1..]
}

fn convert(toml_value: TomlValue) -> JsonValue {
    match toml_value {
        TomlValue::Array(arr) => JsonValue::Array(arr.into_iter().map(convert).collect()),
        TomlValue::Boolean(b) => JsonValue::Bool(b),
        TomlValue::Datetime(dt) => JsonValue::String(dt.to_string()),
        TomlValue::Float(f) => match serde_json::Number::from_f64(f) {
            Some(n) => JsonValue::Number(n),
            None => JsonValue::Null,
        },
        TomlValue::Integer(i) => JsonValue::Number(i.into()),
        TomlValue::String(s) => JsonValue::String(s),
        TomlValue::Table(table) => JsonValue::Object(
            table
                .into_iter()
                .map(|(k, v)| (camel_case(&k), convert(v)))
                .collect(),
        ),
    }
}

fn toml_to_json(source: &str) -> Result<String> {
    let toml = source.parse().context("failed to parse TOML")?;
    let json = serde_json::to_string(&convert(toml)).context("failed serializing JSON")?;
    Ok(json)
}

pub fn get_ui_customization(config: &Config) -> Result<String> {
    let toml_file = &config.ui_customization_file;
    let raw_toml = fs::read_to_string(toml_file)
        .with_context(|| format!("failed reading {}", toml_file.display()))?;
    let ui_customization = toml_to_json(&raw_toml)
        .with_context(|| format!("failed to convert {} to JSON", toml_file.display()))?;
    Ok(ui_customization)
}

#[cfg(test)]
mod tests {
    use std::env;
    use std::fs;
    use std::path::Path;

    use test_case::test_case;

    use super::*;

    #[test_case("oneword" => "oneword")]
    #[test_case("camelCase" => "camelCase")]
    #[test_case("space separated" => "spaceSeparated")]
    #[test_case("underscore_separated" => "underscoreSeparated")]
    #[test_case("multiple   spaces" => "multipleSpaces")]
    #[test_case("   spaces_underscores_mixed" => "spacesUnderscoresMixed")]
    fn camel_case(src: &str) -> String {
        super::camel_case(src)
    }

    #[test]
    fn convert_snapshot() {
        let toml_file = Path::new(file!()).parent().unwrap().join("test.toml");
        let toml_data = fs::read_to_string(toml_file).unwrap();
        insta::assert_yaml_snapshot!(super::convert(toml_data.parse().unwrap()));
    }
}
