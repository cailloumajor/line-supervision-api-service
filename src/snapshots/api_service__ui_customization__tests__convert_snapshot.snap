---
source: src/ui_customization.rs
expression: "super::convert(toml_data.parse().unwrap())"

---
array:
  - val1
  - val2
  - val3
boolean: true
dateTime: "1984-12-09T04:30:00"
floatingPoint: 3.14
infiniteFloat: ~
integer: 42
nanFloat: ~
string: "I'm a string"

